File: 2015/node/day3/twosantas.js
const fs = require('fs');

function moveDirection(direction, x, y) {
    switch (direction) {
        case '<': // Move left
            x -= 1;
            break;
        case '>': // Move right
            x += 1;
            break;
        case '^': // Move up
            y += 1;
            break;
        case 'v': // Move down
            y -= 1;
            break;
        default:
            throw new Error("Invalid direction: " + direction);
    }
    return [x, y];
}

function readFile(filePath) {
    const fileStream = fs.createReadStream(filePath);
    var moves = 0;
    var santa = [0, 0];
    var robosanta = [0, 0];
    var is_santa = moves % 2 === 1;
    let visited = new Set();
    visited.add(JSON.stringify(santa));

    fileStream.on('data', (chunk) => {
        for (let i = 0; i < chunk.length; i++) {
            moves += 1;
            is_santa = moves % 2 == 1;
            const char = String.fromCharCode(chunk[i]); // Convert the byte to a string

            if (is_santa) {
                santa = moveDirection(char, santa[0], santa[1]);
                visited.add(JSON.stringify(santa));
            } else {
                robosanta = moveDirection(char, robosanta[0], robosanta[1]);
                visited.add(JSON.stringify(robosanta));
            }
            
        }
        console.log(`after ${moves} moves, ${visited.size} houses visited`);
    });

    fileStream.on('error', (err) => {
        console.error('Error reading file:', err);
    });
}

const filePath = 'input.txt';
readFile(filePath);

File: 2015/python/day3/twosantas.py
if __name__ == "__main__":
    moves = 0
    santa, robo_santa = (0, 0), (0, 0)
    visited = {(0, 0)}  # Set of visited positions, initialized with (0, 0)

    with open('input.txt') as f:
        for char in f.read():  # Read the entire file at once
            if char not in '<^>v':  # Ignore invalid characters
                continue
            
            moves += 1
            # Determine which character corresponds to which Santa
            is_santa = moves % 2 == 1

            # Update the position based on the direction
            if char == '<':  # Move left
                if is_santa:
                    santa = (santa[0] - 1, santa[1])
                else:
                    robo_santa = (robo_santa[0] - 1, robo_santa[1])
            elif char == '>':  # Move right
                if is_santa:
                    santa = (santa[0] + 1, santa[1])
                else:
                    robo_santa = (robo_santa[0] + 1, robo_santa[1])
            elif char == '^':  # Move up
                if is_santa:
                    santa = (santa[0], santa[1] + 1)
                else:
                    robo_santa = (robo_santa[0], robo_santa[1] + 1)
            elif char == 'v':  # Move down
                if is_santa:
                    santa = (santa[0], santa[1] - 1)
                else:
                    robo_santa = (robo_santa[0], robo_santa[1] - 1)

            # Add the new position to the visited set
            visited.add(santa if is_santa else robo_santa)


    num_visited = len(visited)
    print (f'total houses visited after {moves} moves is {num_visited}')